File: Container_With_Most_Water.py
# -*- coding: utf-8 -*-


class Solution:
    # @return an integer
    def maxArea(self, height):
        if len(height) <= 1:
            return 0
        container_size = 0
        l, r = 0, len(height) - 1
        while l < r:
            if height[l] > height[r]:
                container_size = max(container_size,
                                     height[r] * (r - l))
                r -= 1
            else:
                container_size = max(container_size,
                                     height[l] * (r - l))
                l += 1
        return container_size


if __name__ == '__main__':
    solution = Solution()
    print solution.maxArea([1, 2, 3, 4])

File: src/algorithms/cpp/Container_With_Most_Water.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int maxArea(vector<int>& height) {
        int l = 0, r = SZ(height) - 1;
        int result = (r - l) * min(height[l], height[r]);
        while (l < r) {
            if (height[l] > height[r]) {
                int tmp = (r - l) * height[r];
                result = max(result, tmp);
                r -= 1;
            }
            else {
                int tmp = (r - l) * height[l];
                result = max(result, tmp);
                l += 1;
            }
        }
        return result;
    }
};


int main() {
    Solution solution = Solution();
    vector <int> v;
    v.push_back(5), v.push_back(2), v.push_back(12);
    v.push_back(1), v.push_back(5), v.push_back(3);
    v.push_back(4), v.push_back(11), v.push_back(9);
    v.push_back(4);
    cout << solution.maxArea(v) << endl;
    return 0;
}
